File: src/python/yalix/environment.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class Env(object):

    def __init__(self, local_stack=None, global_frame=None):
        self.local_stack = local_stack if local_stack else list()
        self.global_frame = global_frame if global_frame else dict()
        self.lvar = set(name for name, _ in self.local_stack)

    def extend(self, name, value):
        """
        Extend the local stack with the given name/value.
        Note 2: global frame is shared across extended environments.
        """

        # Prune any shadow bindings, before pushing new name/value
        new_stack = [pair for pair in self.local_stack if pair[0] != name]
        new_stack.append((name, value))

        return Env(local_stack=new_stack,
                   global_frame=self.global_frame)

    def set_local(self, name, value):
        """
        Traverses the local stack and sets the first instance of name with value
        """
        if name in self.lvar:
            stack = self.local_stack
            for i in range(1, len(stack) + 1):
                if self.local_stack[-i][0] == name:
                    self.local_stack[-i] = (name, value)
                    return

        raise ValueError('Assignment disallowed: \'{0}\' is unbound in local environment'.format(name))

    def __getitem__(self, name):
        """
        Look in the local stack first for the named item, then try the global frame
        """
        if name in self.lvar:
            stack = self.local_stack
            while stack:
                peek = stack[-1]
                if peek[0] == name:
                    return peek[1]
                else:
                    stack = stack[:-1]

        if name not in self.global_frame:
            raise ValueError('\'{0}\' is unbound in environment'.format(name))

        return self.global_frame[name]

    def __setitem__(self, name, value):
        """
        Adds a new global definition, and evaluates it according to self
        """
        self.global_frame[name] = value

    def items(self):
        return self.global_frame.items()

File: golang/cmd/environment/environment.go
package environment

import (
	"github.com/pkg/errors"
)

type stackEntry[T any] struct {
	name  string
	value T
}

type Env[T any] struct {
	localStack  []stackEntry[T]
	globalFrame map[string]T
	stackDepth  int
}

func MakeEnv[T any]() Env[T] {
	return Env[T]{
		localStack:  make([]stackEntry[T], 0),
		globalFrame: make(map[string]T),
		stackDepth:  0,
	}
}

// Extend the local stack with the given name/value.
// Note 2: global frame is shared across extended environments.
func (env Env[T]) Extend(name string, value T) Env[T] {
	// Prune any shadow bindings, before pushing new name/value
	var newStack = make([]stackEntry[T], 0)
	for _, entry := range env.localStack {
		if entry.name != name {
			newStack = append(newStack, entry)
		}
	}
	newStack = append(newStack, stackEntry[T]{name, value})

	return Env[T]{
		localStack:  newStack,
		globalFrame: env.globalFrame,
		stackDepth:  env.stackDepth,
	}
}

// Adds a new global definition, and evaluates it according to self
func (env Env[T]) SetGlobal(name string, value T) {
	env.globalFrame[name] = value
}

// Look in the local stack first for the named item, then try the global frame
func (env Env[T]) Get(name string) (T, error) {
	last := len(env.localStack) - 1
	for idx := range env.localStack {
		if env.localStack[last-idx].name == name {
			return env.localStack[last-idx].value, nil
		}
	}

	value, ok := env.globalFrame[name]
	if !ok {
		return value, errors.Errorf("'%s' is unbound in environment", name)
	}

	return value, nil
}

// Traverses the local stack and sets the first instance of name with value
func (env Env[T]) SetLocal(name string, value T) error {
	last := len(env.localStack) - 1
	for idx := range env.localStack {
		if env.localStack[last-idx].name == name {
			env.localStack[last-idx].value = value
			return nil
		}
	}
	return errors.Errorf("assignment disallowed: '%s' is unbound in local environment", name)
}

func (env Env[T]) Includes(name string) bool {
	if _, ok := env.globalFrame[name]; ok {
		return true
	}

	for _, stackEntry := range env.localStack {
		if stackEntry.name == name {
			return true
		}
	}

	return false
}

func (env Env[T]) StackDepth() int {
	return env.stackDepth
}
